crml/nft/src/types.rs

```diff
@@ -0,0 +1,233 @@
+	/// attribute is a timestamp (unix)
+	Timestamp(u64),
+	/// attribute is a stringified URL
+	Url(Vec<u8>),
+}
+
+impl NFTAttributeValue {
+	/// Return the type ID of this attribute value
+	pub const fn type_id(&self) -> NFTAttributeTypeId {
+		match self {
+			NFTAttributeValue::I32(_) => 0,
+			NFTAttributeValue::U8(_) => 1,
+			NFTAttributeValue::U16(_) => 2,
+			NFTAttributeValue::U32(_) => 3,
+			NFTAttributeValue::U64(_) => 4,
+			NFTAttributeValue::U128(_) => 5,
+			NFTAttributeValue::Bytes32(_) => 6,
+			NFTAttributeValue::Bytes(_) => 7,
+			NFTAttributeValue::String(_) => 8,
+			NFTAttributeValue::Hash(_) => 9,
+			NFTAttributeValue::Timestamp(_) => 10,
+			NFTAttributeValue::Url(_) => 11,
+		}
+	}
+	/// Return whether the given `type_id` is valid to describe an `NFTAttribute`
+	pub const fn is_valid_type_id(type_id: NFTAttributeTypeId) -> bool {
+		type_id < (Self::VARIANT_COUNT as u8)
+	}
+	/// Return the byte length of the attribute value, if it exists
+	pub fn len(&self) -> usize {
+		match self {
+			NFTAttributeValue::I32(_) => 4,
+			NFTAttributeValue::U8(_) => 1,
+			NFTAttributeValue::U16(_) => 2,
+			NFTAttributeValue::U32(_) => 4,
+			NFTAttributeValue::U64(_) => 8,
+			NFTAttributeValue::U128(_) => 16,
+			NFTAttributeValue::Bytes32(_) => 32,
+			NFTAttributeValue::Bytes(b) => b.len(),
+			NFTAttributeValue::String(s) => s.len(),
+			NFTAttributeValue::Hash(_) => 32,
+			NFTAttributeValue::Timestamp(_) => 8,
+			NFTAttributeValue::Url(u) => u.len(),
+		}
+	}
+}
+
+/// The max. number of entitlements any royalties schedule can have
+/// just a sensible upper bound
+pub(crate) const MAX_ENTITLEMENTS: usize = 8;
+
+/// Reasons for an auction closure
+#[derive(Decode, Encode, Debug, Clone, PartialEq, Eq)]
+pub enum AuctionClosureReason {
+	/// Auction expired with no bids
+	ExpiredNoBids,
+	/// Auction should have happened but settlement failed due to payment issues
+	SettlementFailed,
+	/// Auction was cancelled by the vendor
+	VendorCancelled,
+}
+
+/// Describes the royalty scheme for secondary sales for an NFT collection/token
+#[derive(Default, Debug, Clone, Encode, Decode, PartialEq)]
+pub struct RoyaltiesSchedule<AccountId> {
+	/// Entitlements on all secondary sales, (beneficiary, % of sale price)
+	pub entitlements: Vec<(AccountId, Permill)>,
+}
+
+impl<AccountId> RoyaltiesSchedule<AccountId> {
+	/// True if entitlements are within valid parameters
+	/// - not overcommitted (> 100%)
+	/// - < MAX_ENTITLEMENTS
+	pub fn validate(&self) -> bool {
+		self.entitlements.is_empty()
+			|| self.entitlements.len() <= MAX_ENTITLEMENTS
+				&& self
+					.entitlements
+					.iter()
+					.map(|(_who, share)| share.deconstruct() as u32)
+					.sum::<u32>() <= Permill::ACCURACY
+	}
+	/// Calculate the total % entitled for royalties
+	/// It will return `0` if the `entitlements` are overcommitted
+	pub fn calculate_total_entitlement(&self) -> Permill {
+		// if royalties are in a strange state
+		if !self.validate() {
```

Comment:
Looks like, we check the validity of royalties on creation of a collection. So this check here can be removed. Also, design wise, the public interfaces shouldn't call each other.